Add parsing support for word-boundary-detection

litherum opened the pull request:
#### 9e639f149ac0564fffed945704406006051f8c6e
<pre>
Add parsing support for word-boundary-detection
<a href="https://bugs.webkit.org/show_bug.cgi?id=257931">https://bugs.webkit.org/show_bug.cgi?id=257931</a>
rdar://110562385

Reviewed by Tim Nguyen.

word-boundary-detection is the mechanism that we can use to expose new linguistic-aware
line breakers to web content. This is a new CSS property which can affect how deeply
browsers perform word detection, which is used for line breaking in languages like
Chinese, Japanese, and Korean. We already have support for these line breakers, but
they&apos;re not currently exposed to the web; this CSS property is the way we can expose
them.

Because we expect to only hook up the `normal` and `auto` values, this patch doesn&apos;t
parse the `manual` value, so that `@supports` will correctly report that we don&apos;t
support that one.

This is tested by tests we&apos;ve already imported from WPT at css/css-text/parsing.

* LayoutTests/imported/w3c/web-platform-tests/css/css-text/parsing/word-boundary-detection-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/parsing/word-boundary-detection-valid-expected.txt:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSValue.cpp:
(WebCore::CSSValue::visitDerived):
* Source/WebCore/css/CSSValue.h:
(WebCore::CSSValue::isWordBoundaryDetectionValue const):
* Source/WebCore/css/CSSWordBoundaryDetectionValue.cpp: Added.
(WebCore::CSSWordBoundaryDetectionValue::CSSWordBoundaryDetectionValue):
(WebCore::CSSWordBoundaryDetectionValue::customCSSText const):
(WebCore::CSSWordBoundaryDetectionValue::equals const):
* Source/WebCore/css/CSSWordBoundaryDetectionValue.h: Added.
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::wordBoundaryDetection):
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeLang):
(WebCore::CSSPropertyParserHelpers::consumeWordBoundaryDetection):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.h:
* Source/WebCore/platform/text/TextFlags.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::rareInheritedDataChangeRequiresLayout):
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::initialWordBoundaryDetection):
(WebCore::RenderStyle::wordBoundaryDetection const):
* Source/WebCore/rendering/style/RenderStyleSetters.h:
(WebCore::RenderStyle::setWordBoundaryDetection):
* Source/WebCore/rendering/style/StyleRareInheritedData.cpp:
* Source/WebCore/rendering/style/StyleRareInheritedData.h:
* Source/WebCore/rendering/style/WordBoundaryDetection.h: Added.
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertWordBoundaryDetection):

Canonical link: <a href="https://commits.webkit.org/265057@main">https://commits.webkit.org/265057@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/105bcf051f4e85ac197847f1ad76048149ff672d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/9654 "6 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/9907 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/10151 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/11316 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/9432 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/9665 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/11890 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/9864 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/12355 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/9807 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/10656 "2 failures") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/8217 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/11475 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/7960 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/8780 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/16183 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/8219 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/9058 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/8928 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/12260 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/9190 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/9415 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/7674 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/9807 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/8601 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/8676 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/2435 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2318 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/12824 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/10067 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/9187 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/2478 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/css/CSSProperties.json:
```diff
@@ -9915,6 +9929,11 @@
+            "shortname": "CSS Fill and Stroke",
+            "longname": "CSS Fill and Stroke Module",
+            "url": "https://www.w3.org/TR/fill-stroke-3/"
+        },
```
Incorrect rebase? I just removed this.

nt1m commented on Source/WebCore/css/CSSProperties.json:
```diff
@@ -6123,6 +6123,20 @@
+                "category": "css-text",
+                "url": "https://drafts.csswg.org/css-text-4/#propdef-word-boundary-detection"
+            }
+        },
```
```suggestion
            },
            status: "in development"
        },
```

nt1m commented on Source/WebCore/css/CSSWordBoundaryDetectionValue.cpp:
You can avoid the need for this class by doing the `WordBoundaryDetection` conversion in `BuilderConverter::convertWordBoundaryDetection`. Up to you

litherum replied:
This class is necessary for serialization. The output of ComputedStyleExtractor is a CSSValue, not a string. So the CSSValue needs to serialize to the right thing. Because the grammar for this property is `normal | manual | auto(<lang>)` it needs a custom serializer.

nt1m reviewed (approved):
This is missing:
* a discrete animation wrapper in CSSPropertyAnimation.cpp and some tests for it (property-list.js), otherwise you'll hit a debug assert.
* a bunch of rebaselines :)

See for a complete list of what you need to change: https://github.com/WebKit/WebKit/commit/9b00aa0f9b088ec6c92453eeb5f76342cfd5eb69

litherum committed:
Add parsing support for word-boundary-detection
https://bugs.webkit.org/show_bug.cgi?id=257931
rdar://110562385

Reviewed by Tim Nguyen.

word-boundary-detection is the mechanism that we can use to expose new linguistic-aware
line breakers to web content. This is a new CSS property which can affect how deeply
browsers perform word detection, which is used for line breaking in languages like
Chinese, Japanese, and Korean. We already have support for these line breakers, but
they're not currently exposed to the web; this CSS property is the way we can expose
them.

Because we expect to only hook up the `normal` and `auto` values, this patch doesn't
parse the `manual` value, so that `@supports` will correctly report that we don't
support that one.

This is tested by tests we've already imported from WPT at css/css-text/parsing.

* LayoutTests/imported/w3c/web-platform-tests/css/css-text/parsing/word-boundary-detection-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/parsing/word-boundary-detection-valid-expected.txt:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSValue.cpp:
(WebCore::CSSValue::visitDerived):
* Source/WebCore/css/CSSValue.h:
(WebCore::CSSValue::isWordBoundaryDetectionValue const):
* Source/WebCore/css/CSSWordBoundaryDetectionValue.cpp: Added.
(WebCore::CSSWordBoundaryDetectionValue::CSSWordBoundaryDetectionValue):
(WebCore::CSSWordBoundaryDetectionValue::customCSSText const):
(WebCore::CSSWordBoundaryDetectionValue::equals const):
* Source/WebCore/css/CSSWordBoundaryDetectionValue.h: Added.
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::wordBoundaryDetection):
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeLang):
(WebCore::CSSPropertyParserHelpers::consumeWordBoundaryDetection):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.h:
* Source/WebCore/platform/text/TextFlags.cpp:
(WebCore::operator<<):
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::rareInheritedDataChangeRequiresLayout):
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::initialWordBoundaryDetection):
(WebCore::RenderStyle::wordBoundaryDetection const):
* Source/WebCore/rendering/style/RenderStyleSetters.h:
(WebCore::RenderStyle::setWordBoundaryDetection):
* Source/WebCore/rendering/style/StyleRareInheritedData.cpp:
* Source/WebCore/rendering/style/StyleRareInheritedData.h:
* Source/WebCore/rendering/style/WordBoundaryDetection.h: Added.
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertWordBoundaryDetection):

Canonical link: https://commits.webkit.org/265057@main
